java/arcs/core/policy/PolicyConstraints.kt

```diff
@@ -132,19 +132,22 @@ sealed class PolicyViolation(val policy: Policy, message: String) : Exception(
     "Policy ${policy.name} violated: $message"
 ) {
     /** Thrown when egress particles were found in the recipe that are not allowed by policy. */
-    class InvalidEgressParticle(
+    class InvalidEgressParticles(
         policy: Policy,
-        val particleNames: List<String>
+        val allowedEgresses: List<String>,
+        val invalidEgresses: List<String>
     ) : PolicyViolation(
         policy,
-        "Egress particle allowed by policy is ${policy.egressParticleName} but found: " +
-            particleNames.joinToString()
+        "Egress particles allowed by policy are " +
+        allowedEgresses.toSortedSet().joinToString(prefix = "{", postfix = "}") +
+        ", but found: " +
+        invalidEgresses.toSortedSet().joinToString(prefix = "{", postfix = "}")
     )
 
-    /** Thrown when multiple egress particles were found in the recipe. */
-    class MultipleEgressParticles(policy: Policy) : PolicyViolation(
+    /** Thrown when policy has no egress particles associated with it. */
+    class PolicyHasNoEgressParticles(policy: Policy) : PolicyViolation(
```

Comment:
This is an interesting one... Is it actually an error if there are no egress particles allowed? That seems like that could actually be a valid state for a policy? I'm thinking about a policy still under development, or maybe one where oyu can use the data with isolated particles in various ways, but no egress.

Reply:
I debated this a bit. Since we are defining policies in terms of validating at egress, I thought it made sense to expect an egress to be always present.  Otherwise, the policy will always be satisfied in the current implementation unless there are already existing checks in the particles. The existing checks may not have anything to do with the policy though. 

I could go either way, but it might be better to start with something strict?

Reply:
That makes sense, sounds good.